torch/csrc/generic/Tensor.cpp

```diff
@@ -568,107 +625,295 @@ static PyObject* THPTensor_(checkAdvancedIndexing)(THPTensor *self, PyObject *ar
+      1 : THTensor_(size)(LIBRARY_STATE indexed.get(), i);
+  }
   THLongTensor *linearIndices = THLongTensor_newWithSize1d(indexingElements);
+
+  // The broadcasted advanced indexing tensor might not be one-dimensional, but we are
+  // generating a vector of indices, so we need to view the indexer as 1D prior to getting
+  // the value for the particular dimension.
+  std::unordered_map<Py_ssize_t, THLongTensorPtr> flattenedBroadcasters;
   THLongStorage *indexerSize = THLongStorage_newWithSize(1);
-  THLongStorage_set(indexerSize, 0, indexingElements);
 
-  for (ptrdiff_t i = 0; i < indexingElements; ++i) {
-    long linearIdx = THTensor_(storageOffset)(LIBRARY_STATE indexed);
-    for (int j = broadcasted.size() - 1; j >= 0; --j) {
-      // Given a series of broadcasted Tensors, we take the jth tensor from the sequence (representing
-      // the indices at dim j) and grab the ith value (used in generating the ith output value)
-      THLongTensor *indexer = THLongTensor_newContiguous(broadcasted.at(j));
-
-      // The indexing tensor might not be one-dimensional, but we are generating a vector of
-      // indices, so we need to view the indexer as 1D prior to getting the value for the
-      // particular dimension
-      THLongTensor *oned = THLongTensor_newView(indexer, indexerSize);
-
-      // Actually laod the value, and verify it is not out-of-bounds
-      long indexAtDim = THLongTensor_get1d(oned, i);
-      long dimSize = THTensor_(size)(LIBRARY_STATE indexed, j);
-      if (indexAtDim >= dimSize) {
-        PyErr_Format(PyExc_IndexError, "index %lld from broadcast indexer is out of range "
-            "for dimension %lld (of size %lld)",
-            (long long)indexAtDim, (long long)j, (long long)dimSize);
-
-        THLongTensor_free(oned);
-        THLongTensor_free(indexer);
-        THLongStorage_free(indexerSize);
-        THLongTensor_free(linearIndices);
-        return NULL;
-      }
+  // All broadcast Tensors have the same number of elements
+  THLongStorage_set(indexerSize,
+                    0,
+                    THLongTensor_nElement(broadcasted.begin()->second.get()));
 
-      linearIdx += THTensor_(stride)(LIBRARY_STATE indexed, j) * THLongTensor_get1d(oned, i);
-      THLongTensor_free(oned);
-      THLongTensor_free(indexer);
-    }
-    THLongTensor_set1d(linearIndices, i, linearIdx);
+  for (auto& broadcast : broadcasted) {
+    THLongTensor *contig = THLongTensor_newContiguous(broadcast.second.get());
+    THLongTensorPtr flat(THLongTensor_newView(contig, indexerSize));
+    flattenedBroadcasters[broadcast.first] = std::move(flat);
+    THLongTensor_free(contig);
   }
   THLongStorage_free(indexerSize);
 
+  long baseOffset = THTensor_(storageOffset)(LIBRARY_STATE indexed);
+  for (ptrdiff_t i = 0; i < indexingElements; ++i) {
+    long linearIdx = THPTensor_(_indexToOffset)(
+        indexed, flattenedBroadcasters, i);
+    THLongTensor_set1d(linearIndices, i, baseOffset + linearIdx);
```

Comment:
you can use `THLongTensor_fastSet1d` which will be much faster, because it's inlined and doesn't do bounds-checks.
Similarly, wherever you used set1d, use THTensor_fastSet1d